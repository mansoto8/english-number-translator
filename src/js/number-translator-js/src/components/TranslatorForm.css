.translator-form {
    display: flex;
    margin-top: 30px;
    margin-bottom: 15px;
}

.result-div {
    display: flex;
    align-items: center;
    text-align: center;
}

#translation-container {
    width: 1200px;
}

#translation-output {
    width: 900px;
}

#translation-output-label {
    padding-top: 5px;
    width: 200px;
    margin-right: 20px;
}
#inputGroupNumber {
    width: 200px;
    margin-right: 20px;
}
