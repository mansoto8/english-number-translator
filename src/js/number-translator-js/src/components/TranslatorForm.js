import React, {Component} from 'react';
import {Button, Container, Form, FormGroup, Input, Label} from 'reactstrap';
import './TranslatorForm.css';

class TranslatorForm extends Component {

    emptyItem = {
        number: 0,
        translation: ''
    };

    constructor(props) {
        super(props);
        this.state = {
            item: this.emptyItem,
        };
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange(event) {
        const target = event.target;
        const value = target.value;
        const name = target.name;
        let item = {...this.state.item};
        item[name] = value;
        this.setState({item});
    }

    async handleSubmit(event) {
        event.preventDefault();
        let {item} = this.state;

        const translation = await fetch('/api/numbers' + (item.number ? '/' + item.number : ''), {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
        });

        const resp = await translation.json();
        item.translation = resp.output;
        this.setState({item});
    }

    render() {
        const {item} = this.state;

        return <div>
            <Container>
                <Form onSubmit={this.handleSubmit} className={"translator-form"}>
                    <Input type="number" name="number" id="number-input" value={item.number || ''}
                           onChange={this.handleChange} placeholder="Enter a number"/>
                    <Button color="primary" type="submit">Translate</Button>{' '}
                </Form>
                <div className="result-div">
                    <Label for="translation-output" id={"translation-output-label"}>The number translated is: </Label>
                    <Input type="text" name="translation-output" id="translation-output" value={item.translation || ''}
                           readOnly/>
                </div>
            </Container>
        </div>
    }
}

export default TranslatorForm;
