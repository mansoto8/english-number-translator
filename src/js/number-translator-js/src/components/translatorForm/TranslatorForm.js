import {Container, FormGroup, Label, FormFeedback, Form, Input, InputGroup, Button} from 'reactstrap';
import React, {Component} from 'react';
import './TranslatorForm.css';

class TranslatorForm extends Component {

    emptyItem = {
        number: undefined,
        translation: '',
    };

    validate = {
        number: 'valid-number-format',
    };

    constructor(props) {
        super(props);
        this.state = {
            item: this.emptyItem,
            validate: this.validate,
        };
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.isValidNumberFormat = this.isValidNumberFormat.bind(this);
    }

    handleChange(event) {
        const target = event.target;
        const value = target.value;
        const name = target.name;
        let item = {...this.state.item};
        item[name] = value;
        this.setState({item});
    }

    async handleSubmit(event) {
        event.preventDefault();
        let {item, validate} = this.state;

        let number = item.number;
        if (!this.isValidNumberFormat(number)) {
            validate.number = 'invalid-number-format';
            this.setState({validate});
            return;
        }

        if (number.indexOf(',') > 0) {
            number = number.replace(/,/g, '');
        }

        if (Math.abs(number) > 9999999999) {
            validate.number = 'invalid-number-size';
            this.setState({validate});
            return;
        }

        const translation = await fetch('/api/numbers/' + number, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
        });

        const resp = await translation.json();
        item.translation = resp.output;
        validate.number = 'valid-number-format';
        this.setState({item, validate});
    }

    isValidNumberFormat(number) {
        if (number === undefined || number === null || number.length === 0) {
            return false;
        }

        const commaIndex = number.indexOf(',');
        if (commaIndex >= 0) {
            const size = number.length;
            let validCommaIndex = size - 4;
            for (let i = size - 1; i >= 0; i--) {
                const character = number.charAt(i);
                if (character === '-') {
                    if (i === 0) {
                        return number.charAt(i + 1) !== ',';
                    }
                }
                if (i === validCommaIndex) {
                    if (character === ',' && i !== 0) {
                        validCommaIndex -= 4;
                    } else {
                        return false;
                    }
                } else {
                    if (isNaN(character)) {
                        return false;
                    }
                }
            }
        }

        return commaIndex >= 0 ? true : !isNaN(number);
    }

    getErrorMessage(validate) {
        let message;

        switch (validate.number) {
            case 'invalid-number-format':
                message = 'Incorrect number format';
                break;
            case  'invalid-number-size':
                message = 'Number is too big. Max allowed is +/- 9,999,999,999';
                break;
            default:
                message = '';
        }

        return message;
    }

    render() {
        const {item, validate} = this.state;

        const errorMessage = this.getErrorMessage(validate);

        return <div>
            <Container>
                <Form onSubmit={this.handleSubmit} className={"translator-form"}>
                    <InputGroup id="inputGroupNumber">
                        <Input type="text" name="number" id="number-input" value={item.number || ''}
                               valid={errorMessage === ''}
                               invalid={errorMessage !== ''}
                               onChange={this.handleChange} placeholder="Enter a number"/>
                        <FormFeedback id="formFeedBackNumber" invalid="true">{errorMessage}</FormFeedback>
                    </InputGroup>
                    <FormGroup>
                        <Button id="translate-button" color="primary" type="submit">Translate</Button>
                    </FormGroup>
                </Form>
                <div className="result-div">
                    <Label for="translation-output" id={"translation-output-label"}>The number translated is: </Label>
                    <Input type="text" name="translation-output" id="translation-output" value={item.translation || ''}
                           readOnly/>
                </div>
            </Container>
        </div>
    }
}

export default TranslatorForm;
