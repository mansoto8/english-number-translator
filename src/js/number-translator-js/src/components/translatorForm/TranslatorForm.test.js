import TranslatorForm from './TranslatorForm';
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import {configure} from 'enzyme';
import {shallow} from 'enzyme';

configure({adapter: new Adapter()});

it('renders initial form', () => {
    const wrapper = shallow(
        <TranslatorForm/>,
    );

    expect(wrapper).toMatchSnapshot();
});

it('validates number format', () => {
    const wrapper = shallow(
        <TranslatorForm/>,
    );
    const instance = wrapper.instance();

    expect(instance.isValidNumberFormat(null)).toBe(false);
    expect(instance.isValidNumberFormat(undefined)).toBe(false);
    expect(instance.isValidNumberFormat("")).toBe(false);

    expect(instance.isValidNumberFormat("words")).toBe(false);
    expect(instance.isValidNumberFormat("wordsAnd13413")).toBe(false);
    expect(instance.isValidNumberFormat("w8r4ik4-,")).toBe(false);
    expect(instance.isValidNumberFormat("1,333,34")).toBe(false);
    expect(instance.isValidNumberFormat(",333,34")).toBe(false);
    expect(instance.isValidNumberFormat("-,333,34")).toBe(false);
    expect(instance.isValidNumberFormat("333-34")).toBe(false);
    expect(instance.isValidNumberFormat("33334-")).toBe(false);
    expect(instance.isValidNumberFormat("33334-")).toBe(false);
    expect(instance.isValidNumberFormat(",")).toBe(false);
    expect(instance.isValidNumberFormat("-,")).toBe(false);
    expect(instance.isValidNumberFormat("-")).toBe(false);
    expect(instance.isValidNumberFormat("000010")).toBe(true);
    expect(instance.isValidNumberFormat("-0010")).toBe(true);
    expect(instance.isValidNumberFormat("0000")).toBe(true);
    expect(instance.isValidNumberFormat("-20")).toBe(true);
    expect(instance.isValidNumberFormat("200")).toBe(true);
});
