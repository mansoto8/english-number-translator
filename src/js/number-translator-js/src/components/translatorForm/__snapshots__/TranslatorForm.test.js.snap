// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders initial form 1`] = `
<div>
  <Container
    tag="div"
  >
    <Form
      className="translator-form"
      onSubmit={[Function]}
      tag="form"
    >
      <InputGroup
        id="inputGroupNumber"
        tag="div"
      >
        <Input
          id="number-input"
          invalid={false}
          name="number"
          onChange={[Function]}
          placeholder="Enter a number"
          type="text"
          valid={true}
          value=""
        />
        <FormFeedback
          id="formFeedBackNumber"
          invalid="true"
          tag="div"
        />
      </InputGroup>
      <FormGroup
        tag="div"
      >
        <Button
          color="primary"
          id="translate-button"
          tag="button"
          type="submit"
        >
          Translate
        </Button>
      </FormGroup>
    </Form>
    <div
      className="result-div"
    >
      <Label
        for="translation-output"
        id="translation-output-label"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        The number translated is: 
      </Label>
      <Input
        id="translation-output"
        name="translation-output"
        readOnly={true}
        type="text"
        value=""
      />
    </div>
  </Container>
</div>
`;
